062-topic-replication-factor-change.md

```diff
@@ -0,0 +1,223 @@
+
+The topic configuration endpoint is asynchronous, which means that it validates the task, but the execution will be queued.
+Inside Cruise Control, a valid topic configuration task is first create in an `Active` state, then moved to `InExecution`, and finally to `Completed` or `CompletedWithError`.
+Only a single task can be `InExecution` state at any given time, which includes both topic configuration and rebalance tasks.
+
+When Cruise Control integration is enabled, a `ReplicationFactorManager` is created to manage the communication with the REST API.
+The `ReplicationFactorManager` stores all ongoing replication factor changes in an in-memory cache, in order to periodically check their status.
+A single replication factor change can be ongoing at any given time for a specific `KafkaTopic` resource, following replication factor changes will remain pending, and periodically retried.
+At most `max.active.user.tasks` (default: 5) replication factor changes for different `KafkaTopic` resources can be ongoing at the same time.
+
+Cruise Control allows to scale down the replication factor under the `min.insync.replicas` value, and this can cause an availability issue for producers with `acks=all`.
+We won't implement any check for that, because the `Cluster Operator` ignores pathological topic configurations where RF < minISR, and these topics don't event show up as under replicated.
+The target replication factor should also be less than or equal to the number of available brokers, but this is enforced directly by Cruise Control.
+
+Even if the replication factor change cannot be executed for some reason, the `KafkaTopic` remains available.
+All Cruise Control errors are appended as warnings in `KafkaTopic` status, but logged with the appropriate error level.
+
+Cruise Control exceptions:
+
+- `NotEnoughValidWindowsException` can be raised if the cluster workload model is not ready when a replication factor change is requested.
+  This may happen when Cruise Control has just started and does not have enough samples, or when there is some issue in producing and consuming metrics.
+  In this case, the Topic Operator replaces Cruise Control's error message with a custom warning message.
+- `OngoingExecutionException` can be raised if there is already a task that is being executed.
+  In this case, the Topic Operator replaces Cruise Control's error message with a custom warning message.
+- `OptimizationFailureException` can be raised if the target replication factor is smaller than the number of available brokers, or hard goals are violated.
+  In both cases the Topic Operator will continue to retry on every periodic reconciliation, until `KafkaTopic` configuration is updated, or the environment matches the desired state.
+  For example, a temporary `NetworkInboundCapacityGoal` violation can be caused by a load test where we send a lot of data in a very short period of time.
```

Comment:
This sentence seems somehow out of place? I'm not sure I follow how it relates, what error does it produce, how will the operator handle it.

Reply:
In this case, it just waits for the inbound capacity to settle down, and then it will apply the RF change as usual. Instead, if the violation is permanent, it will keep the status warning.